rfd/0000-prototype-job-worker.md

```diff
@@ -0,0 +1,203 @@
+    interacting with processes in other namespaces.
+  - **Network Namespace**: Isolates network interfaces, IP routing tables, port
+    numbers, etc., allowing processes to operate on a network as if they were
+    the only ones on the system. To stay in scope, the network will remain
+    isolated.
+  - **Mount Namespace**: Isolates filesystem mount points, allowing processes
+    to have their own set of filesystems visible to them.
+
+##### Utilization Strategy
+Our service will utilize the syscall package in Go to unshare and manage
+namespaces. The following code snippet demonstrates how a new process can be
+started in a new set of namespaces:
+
+```go
+import "syscall"
+
+func setupNamespaces() error {
+    // Unshare creates new namespaces for the calling process
+    if err := syscall.Unshare(syscall.CLONE_NEWPID | syscall.CLONE_NEWNET | syscall.CLONE_NEWNS); err != nil {
+        return fmt.Errorf("failed to create namespaces: %v", err)
+    }
+    return nil
+}
+```
+
+#### Cgroup Implementation
+Control groups (cgroups) are another kernel feature that limits, accounts for,
+and isolates the resource usage (CPU, memory, disk I/O, etc.) of process
+groups. We will implement cgroups to ensure that each job runs within
+predefined resource limits, improving system stability and preventing any job
+from monopolizing system resources.
+
+##### Configuration Approach
+Cgroups will be managed by interacting with the cgroup filesystem typically
+mounted under /sys/fs/cgroup. Here is how we plan to create and manage cgroups:
+
+```go
+import (
+    "io/ioutil"
+    "path/filepath"
+)
+
+func createCgroup(path, resource, limit string) error {
+    cgroupPath := filepath.Join("/sys/fs/cgroup", resource, path)
+    if err := os.MkdirAll(cgroupPath, 0755); err != nil {
+        return err
+    }
+    return ioutil.WriteFile(filepath.Join(cgroupPath, "tasks"), []byte(strconv.Itoa(os.Getpid())), 0644)
+}
+```
+
+### Command Usage
+
+```
+Usage: jobworker [command] [options]
+
+A CLI tool to interact with the Job Worker Service, providing secure job management capabilities through mTLS.
+
+Commands:
+  start       Start a new job with the specified command and arguments.
+  stop        Stop a running job using its unique job ID.
+  status      Query the status of a job using its unique job ID.
+  stream      Stream the output of a running job using its unique job ID.
+  run         Run a local process in specified namespaces and cgroup.
+
+Options:
+  --help      Show this help message and exit.
+  --cert      Specify the client certificate (and optionally key) file path for mTLS authentication.
+              Defaults to ~/.ssl/client.pem if not specified.
+
+Examples:
+  jobworker start --cmd "bash myscript.sh"                          Start a job using default certificate and key.
+  jobworker stop --job-id 12345 --cert /path/to/custom.pem          Stop a job using a specified PEM file.
```

Comment:
How would job id be generated?

Reply:
Since there wont be any persistence between runs, they will be sequential, and restart at `0` on each run.